### PR TITLE
Add support for extended attributes and quick attribution of DMGs

### PR DESCRIPTION
**Pull request by bhearsum**
This PR includes everything that is needed to make macOS DMGs attributable. That is:
- Support for adding extended attributes to HFS+ filesystems with `hfsplus setattr`
- Support for making a DMG quickly attributable during `dmg build`
- Support for updating attribution information with the new `dmg attribute` command

Some rough edges I'd like to call out:
- There's a TODO about checking partition ranges. I'm honestly not sure if this is necessary or not. @ncalexan - can you provide a bit more context on this, if you have it?
- If multiple files need to have extended attributes on them, they _must_ be added in order of their `fileID`. This is unfortunate, but does not interfere with our plans (we only need one extended attribute). I suspect it's theoretically possible to fix this by reorganizing the BTree while adding additional attributes - but I'm honestly not certain about this, nor do I think it's worth the effort right now.
- The writing out of the attribution resource in `resources.c` is pretty nasty -- using a flag to signal whether or not to treat it as binary data or not. I attempted to make this cleaner by storing the `plst` name as an already serialized string (so that the existing code would write it correctly), but I couldn't quite get this right, and it ended up moving the base64 encoding & flipping into places like `buildDmg` - which felt just as icky.

And some other notes:
- The `run_spanning` tests were very tricky to construct - they rely on setting a large number of extra extended attributes to push the one containing the sentinel past the boundaries of a single `run`. I've included the script I used to generate these extra attributes as well as the exact list that forces this to happen. (Despite setting them all first, some end up after the sentinel attribute, for reasons I couldn't decipher.)
- @ncalexan's work that this is based on had a TODO about sectors aligning badly. I found that each `run` that `io.c` processes reads all sectors in one pass - so I don't think this is necessary to handle. However, this led me to noticing that there are multiple `runs` that we might cross (maybe this is what you meant by sectors?) which is now handled correctly.
- At one point, I believe the plan was to pad the attribution sentinel to 1k by filling it with nul bytes. This does not work, unfortunately (when the extended attribute values are read back, they stop at the first nul byte, even if many were written). Instead, I intend for us to pad with tabs or newlines (this should make things easier on parsers - which can just strip trailing whitespace before parsing). It's possible I'm overlooking or missing something here - so if there's a better idea please do raise it.

@ncalexan - I know you probably can't officially review this whole patch seeing as you wrote a large portion of it, but given some of the changes I've made I think you ought to weigh in at least.

@glandium - you seem like the most likely person to review this. I can't seem to official mark you as a reviewer though, so hopefully you see this!

### CODE REVIEW

**Comment by glandium** [comment]
Before I dive into this, I'll say this: the current tip of this repo doesn't work for mozilla-central. See [bug 1790728](https://bugzilla.mozilla.org/show_bug.cgi?id=1790728).

**Comment by glandium** [comment]
@gabrielesvelto might want to chime in, too.

**Comment by bhearsum** [comment]
> Before I dive into this, I'll say this: the current tip of this repo doesn't work for mozilla-central. See [bug 1790728](https://bugzilla.mozilla.org/show_bug.cgi?id=1790728).

Thanks for the heads up! Depending on whether or not someone plans to look at that soon can we consider backing it out? (Or I could land this work on a separate branch from an older commit...)

**Comment by gabrielesvelto** [comment]
We can back out the problematic patches on the current tip. I've wanted to look into the issues for the macOS system libraries symbol scrapers for a long while but never prioritized it because of a combination of other stuff to do and Steven Michaud sending us the symbols he scrapes manually anyway. That's a long way to say I won't be able to fix the existing issues soon so better get rid of them.

**Comment by bhearsum** [comment]
@ncalexan, @glandium, @gabrielesvelto - I'm not sure who wants to / should review this - so I'm flagging a few of you - feel free to remove yourself if you don't think you're an appropriate reviewer.

(This does need rebasing, but only because of the backout in #7 - nothing noteworthy will change after rebasing.)

File: dmg/attribution.c [hunk 1]
```diff
@@ -42,10 +42,23 @@ FindStrInBuf(char const * buf, size_t bufLen, char const * str)
...
-	return NULL != FindStrInBuf((char const*)data, len, (const char*)attributionData->sentinel);
+	if (NULL != FindStrInBuf((char const*)data, len, (const char*)attributionData->sentinel)) {
+	  return KeepCurrentRaw;
+	}
+
+	// If we didn't find it in the data, check if it spans data + nextData
+	if (nextLen > 0) {
+	  char* combinedData = malloc(len + nextLen + 1);
+	  memcpy(combinedData, data, len);
+	  memcpy(combinedData + len, nextData, nextLen);
```
Comment (gabrielesvelto) [inline dmg/attribution.c]: We're allocating `len + nextLen + 1` bytes but only setting `len + nextLen`, I think we might be missing a nul termination? Or maybe drop the +1 and just pass `len + nextLen` to `FindStrInBuf()`?

**Review by gabrielesvelto — CHANGES_REQUESTED** [review]
This generally looks good but my knowledge of this code is limited so there's that. I've found only one issue that should be addressed but the rest looks solid.

File: dmg/attribution.c (hunk 1 above)
Comment (bhearsum) [reply dmg/attribution.c]: Good catch - thank you! I will update this one way or the other.

**Comment by bhearsum** [comment]
This has now been rebased on top of the `mozilla` branch. In addition to fixing the potential out of bounds you noted, I also fixed a couple of minor things I noticed during additional testing:
- zeroing out the attribution area now works even if only a prefix of the sentinel is passed to `dmg attribute`
- the `signature` field of the `AttributionResource` is now endian flipped for reading/writing like the rest of the structure

(https://github.com/mozilla/libdmg-hfsplus/compare/5bbc90f8b1e525b399860f03e10329c14c6c5b9c..0950317f39576a6506e62f1772435a86554a11b2 contains a diff of the previously reviewed changes against the latest push for easier reviewing.)

**Review by gabrielesvelto — APPROVED** [review]
Looking good, let's go with this

**Pull request merged**